js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package grpc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/dop251/goja"
+	//nolint: staticcheck
+	"github.com/golang/protobuf/proto"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/desc/protoparse"
+	"github.com/jhump/protoreflect/dynamic"
+	"github.com/jhump/protoreflect/dynamic/grpcdynamic"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	grpcstats "google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+
+	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/metrics"
+	"github.com/loadimpact/k6/stats"
+)
+
+//nolint: lll
+var (
+	errInvokeRPCInInitContext = common.NewInitContextError("invoking RPC methods in the init context is not supported")
+	errConnectInInitContext   = common.NewInitContextError("connecting to a gRPC server in the init context is not supported")
+)
+
+// Client reprecents a gRPC client that can be used to make RPC requests
+type Client struct {
+	mds  map[string]*desc.MethodDescriptor
+	tags map[string]string
+	conn *grpc.ClientConn
+}
+
+// NewClient creates a new gPRC client to make invoke RPC methods.
+func (*GRPC) NewClient(ctxPtr *context.Context) interface{} {
+	rt := common.GetRuntime(*ctxPtr)
+
+	return common.Bind(rt, &Client{}, ctxPtr)
+}
+
+// MethodInfo holds ifromation on any parsed method descriptors that can be used by the goja VM
+type MethodInfo struct {
+	FullMethod     string
+	IsClientStream bool
+	IsServerStream bool
+}
+
+// Response is a gRPC response that can be used by the goja VM
+type Response struct {
+	Status   codes.Code
+	Message  goja.Value
+	Headers  map[string][]string
+	Trailers map[string][]string
+	Error    goja.Value
+}
+
+// Load will parse the given proto files and make the file descriptors available to request.
+func (c *Client) Load(ctxPtr *context.Context, importPaths []string, filenames ...string) ([]MethodInfo, error) {
+	if lib.GetState(*ctxPtr) != nil {
+		return nil, errors.New("load must be called in the init context")
+	}
+
+	parser := protoparse.Parser{
+		ImportPaths:      importPaths,
+		InferImportPaths: len(importPaths) == 0,
+	}
+
+	fds, err := parser.ParseFiles(filenames...)
+	if err != nil {
+		return nil, err
+	}
+
+	var rtn []MethodInfo
+	c.mds = make(map[string]*desc.MethodDescriptor)
+	for _, fd := range fds {
+		for _, sd := range fd.GetServices() {
+			for _, md := range sd.GetMethods() {
+				var s strings.Builder
+				s.WriteString(sd.GetFullyQualifiedName())
+				s.WriteRune('/')
+				s.WriteString(md.GetName())
+				name := s.String()
+				c.mds[name] = md
```

Comment:
That would be impossible, as this would mean you had the exact same same package, service and method defined more than once, which would mean your protocol buffer definition would fail to compile in the first place.

Reply:
hmm true, but here we can load a bunch of `.proto` files, couldn't conflicts exist between them? 
```js
let client = grpc.newClient();
client.load([], "some-definition.proto")
client.load([], "other-definition-with-conflicts.proto")
```

Reply:
One load will override the other load; so in your example only `other-definition-with-conflicts.proto` would be loaded.

You would more likely load with multiple in one load statement: 
```proto
client.load([], "one.proto", "two.proto", "etc.proto")
```
In this case all definitions and their dependencies are validated before we iterate over the services and methods; so if there was a duplicate it would be registered more than once, which would result in an early error; I already check for duplicate file registrations as part of the parsing of dependancy imports.

I can still do a check in this map to, and log a warning (?), but I'm guessing it would never get there (server implementation would fail if this was valid).

Let me test the theory to see what happens.